### PR TITLE
Improve support for gen2 and gen3 tags

### PR DESCRIPTION
**Pull request by gelotus**
Ok, my ocd wins..
There is a rework including @AdamLaurie commits.
Tested before and after on genuine and gen2. 
Now you can rewrite gen2 with W, please see commit descs

### CODE REVIEW

**Comment by AdamLaurie** [comment]
Nice work @gelotus - thank you! But please modify unlock_card() so it does not write on READ commands. If this is the only way to detect Magic cards then we should print a warning that we are about to try to write to the card and ask if the user wants to go ahead or skip that test.

**Comment by gelotus** [comment]
Ok, fixed. About your second question, about one month ago I've bought a lot of 500 gen2 tags,for what I see, they have a random uid from factory. I haven't checked rigorously, but from my memory I remember so. Maybe when I'll buy another batch I'll check better. If we use is_directwrite() we break gen3 so the tag burn during check. 
About the warning and the user intervention, this will break a possible scripted using of the program, for example I'm using it for quickly format that 500 tags what I have talked before. We can add an override switch but I think would be better to rework the args parser first. Anyway I don't see troubles whit this behaviour, if we try to W to a genuine or gen3 tag, writing simply fails, for gen1 it continues with its normal behaviour. If you think about a possible overwrite of the content, the same happens on a gen1.

File: utils/nfc-mfclassic.c [hunk 1]
```diff
@@ -587,6 +587,7 @@ print_usage(const char *pcProgramName)
   printf("                  *** unlocked read does not require authentication and will reveal A and B keys\n");
   printf("                  *** note that unlocked write will attempt to overwrite block 0 including UID\n");
   printf("                  *** unlocking only works with special Mifare 1K cards (Chinese clones)\n");
+  printf("                  *** unlocking (w) also works for rewriting block 0 in gen2 clones\n");
```
Comment (AdamLaurie) [inline utils/nfc-mfclassic.c:585]: should be (W) ?

Comment (gelotus) [reply utils/nfc-mfclassic.c:585]: Yes. With a uppercase W, we set the write_block_zero to true, that goes to unlock_card(), that try to sends unlock command for a gen1 tag, if the first unlock command fails, set magic2 to true so in the end sets writing to block 0 with auth. If unlocks commands don't fails its goes to normal gen1 behaviour so writing all blocks including 0 without auth. With a lowercase w all goes like before. All the reading process behaviour haven't changed (at least with that read-write check in unlock_card()).

**Comment by gelotus** [comment]
Sorry, I hadn't noticed that typo..

**Comment by AdamLaurie** [comment]
Note that I have merged this as the non-writing of full sectors is a bigger issue but there is still a problem here with 4k DirectWrite - on a read operation standard 4K cards are being recognised as DirectWrite.

**Comment by gelotus** [comment]
Yes I'll could take a look , but unfortunately I haven't a 4k tag for testing. We'll see in the next days if I'll find one

**Pull request merged**